server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -3914,6 +3949,7 @@ public InternalEngine recoverFromTranslog(TranslogRecoveryRunner translogRecover
         shard.acquireAllReplicaOperationsPermits(shard.getOperationPrimaryTerm(), shard.getLastKnownGlobalCheckpoint(), 0L,
             ActionListener.wrap(r -> {
                 try (r) {
+                    resetWasStarted.set(true);
```

Comment:
It looks like we rely on timing to hit the cases where reset has started and has not started. I think I would prefer explicit latches/barriers to more directly dictate the order?